### ISSUE TITLE
TimeDistributed CNN + LSTM masking not working

### ISSUE THREAD

**Issue by tinder-arozinov** [keras-team/keras#10028]
I'm trying to implement an architecture similar to this one: 

    Time = 0    Time = 1    ....    Time = t
      im@0         im@1     ....       im@t
        |            |                                |
       CNN      CNN        ....        CNN
        |            |                                 |
       RNN--------->RNN---->....------->RNN
                                         |
                                   some prediction 

The associated code is (for tensorflow backend):

```
MAX_CROPS=4
SIZE_RET=(299,299)
IMAGE_MASK= -2*np.ones((1,SIZE_RET[0],SIZE_RET[1],3))

base_model = inception_v3.InceptionV3(weights='imagenet', include_top=False)
x = base_model.output
x = GlobalAveragePooling2D()(x)
x = Dense(2048,activation="relu")(x)
base_model = Model(base_model.input,x)

#output of CNN on each crop
CNN_OUTPUT_SIZE = int(base_model.output.shape[-1])
print(CNN_OUTPUT_SIZE)

LSTM_MASK = np.zeros((1,CNN_OUTPUT_SIZE))

input_layer = Input(shape=(MAX_CROPS, SIZE_RET[0], SIZE_RET[1], 3))
#Use dummy image to mask non-body-crops
curr_layer= Masking(mask_value = IMAGE_MASK)(input_layer)

#freeze tops of CNN
for layer in base_model.layers[0:41]:
   layer.trainable =False

#Make cnn predict on each crop
curr_layer = TimeDistributed(base_model)(curr_layer)

#Reshape doesn't support masking for some reason, so need a dummy identity layer to fix this??
curr_layer=Lambda(lambda x: x, output_shape=lambda s:s)(curr_layer)
curr_layer = Reshape(target_shape=(MAX_CROPS, CNN_OUTPUT_SIZE))(curr_layer)

#Second Mask for LSTM
curr_layer = Masking(mask_value = LSTM_MASK)(curr_layer)
lstm_out = LSTM(LSTM_OUTPUT_DIM,return_sequences=False)(curr_layer)
#lstm_out = Dense(FC_DIM)(lstm_out)

lstm_out=Model(input_layer,lstm_out)
```

I'm suspecting that the masking is not working correctly here. What seems to be happing is that the first masking layer prior to the CNN correctly identifies the input image mask, and sets it to all 0s. This is already problematic because the CNN now assumes I'm putting in an all-black image (disregard preprocessing here). This goes on to be passed to the CNN (which I specifically don't want), which outputs junk values that are nonzero. That's why i set the LSTM_MASK to all 0s, because that's what I would like the output of the CNN to be. Maybe I'm not understanding the masking layer here correctly?

What can I do to ensure the CNN outputs a dummy mask on an input that has masks?

**Comment by sim-san** [comment]
@tinder-arozinov  Have you found a solution?

**Comment by ririya** [comment]
I solved the problem by adding a second input, a mask to specify which timesteps to take into account. That way the image sequence always has the same number of timesteps, the CNN always generates an output, but some of them are ignored for the LSTM input. However, the missing images need to be chosen carefully so that batch normalization is not affected.

def LSTM_CNN(params):

resnet = ResNet50(include_top=False, weights='imagenet', pooling = 'avg')

input_layer = Input(shape=(params.numFrames, params.height, params.width, 3))

input_mask = Input(shape=(params.numFrames,1))

curr_layer = TimeDistributed(resnet)(input_layer)

resnetOutput = Dropout(0.5)(curr_layer)

curr_layer = multiply([resnetOutput,input_mask])

cnn_output = curr_layer

curr_layer = Masking(mask_value=0.0)(curr_layer)

lstm_out = LSTM(256, dropout=0.5)(curr_layer)
output = Dense(output_dim=params.numClasses, activation='sigmoid')(lstm_out)

model = Model([input_layer, input_mask], output)

return model
